system_metrics_collector/src/topic_statistics_collector/subscriber_topic_statistics.hpp

```diff
@@ -160,10 +219,165 @@ class SubscriberTopicStatisticsNode : public system_metrics_collector::MetricsMe
+    publish_param_desc);
+  publish_period_ = std::chrono::milliseconds{publish_period};
+
+  const auto collect_topic_desc = buildTopicParameterDescriptor(
+    "The topic to subscribe to, for calculating topic statistics.");
+  collect_topic_name_ = declare_parameter(
+    topic_statistics_constants::kCollectStatsTopicNameParam,
+    topic_statistics_constants::kDefaultCollectStatsTopicName,
+    collect_topic_desc);
+  validateStringParam(collect_topic_name_);
+
+  const auto publish_topic_desc = buildTopicParameterDescriptor(
+    "The topic to publish topic statistics to.");
+  publish_topic_name_ = declare_parameter(
+    topic_statistics_constants::kPublishStatsTopicNameParam,
+    system_metrics_collector::collector_node_constants::kStatisticsTopicName,
+    publish_topic_desc);
+  validateStringParam(publish_topic_name_);
+
+  initializeCollectors(statistics_collectors_);
+
+  auto callback = [this](typename T::SharedPtr received_message) {
+      for (const auto & collector : statistics_collectors_) {
+        collector->OnMessageReceived(*received_message, this->LifecycleNode::now().nanoseconds());
+      }
+    };
+  subscription_ = create_subscription<T>(collect_topic_name_, 10, callback);
+}
+
+template<typename T>
+void SubscriberTopicStatisticsNode<T>::StartPublisher()
+{
+  rcpputils::check_true(publish_timer_ == nullptr);
+
+  if (publisher_ == nullptr) {
+    publisher_ = create_publisher<MetricsMessage>(publish_topic_name_, 10);
+  }
+
+  publisher_->on_activate();
+
+  RCLCPP_DEBUG(this->get_logger(), "SetupStart: creating publish_timer_");
+  publish_timer_ = this->create_wall_timer(
+    publish_period_, [this]() {
+      this->PublishStatisticMessage();
+      this->ClearCollectorMeasurements();
+    });
+  window_start_ = now();
+}
+
+template<typename T>
+rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn
+SubscriberTopicStatisticsNode<T>::on_activate(const rclcpp_lifecycle::State &)
+{
+  RCLCPP_DEBUG(this->get_logger(), "on_activate");
+  for (const auto & collector : statistics_collectors_) {
+    if (!collector->Start()) {
+      return CallbackReturn::ERROR;
+    }
+  }
+
+  StartPublisher();
+  return CallbackReturn::SUCCESS;
+}
+
+template<typename T>
+rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn
+SubscriberTopicStatisticsNode<T>::on_deactivate(const rclcpp_lifecycle::State & state)
+{
+  RCLCPP_DEBUG(this->get_logger(), "on_deactivate");
+  for (const auto & collector : statistics_collectors_) {
+    if (!collector->Stop()) {
+      return CallbackReturn::ERROR;
+    }
+  }
+
+  StopPublisher();
+  return CallbackReturn::SUCCESS;
+}
+
+template<typename T>
+rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn
+SubscriberTopicStatisticsNode<T>::on_shutdown(const rclcpp_lifecycle::State & state)
+{
+  RCLCPP_DEBUG(this->get_logger(), "on_shutdown");
+  ClearCollectorMeasurements();
```

Comment:
Don't you need to call stop? If you can, mirror PeriodicMeasurementNode and we'll accept the tech debt to merge the commonalities.

Note: check the ros2 lifecycle transitions. Something can be shutdown or on error before deactivate is called. You can verify these changes with unit tests as well.

Reply:
Yes, `StopPublisher()` needed to be called here and in shutdown (similar to `PeriodicMeasurementNode`).

It may not be easy to merge commonalities with `PeriodicMeasurementNode` because it depends on `Collector::Stop()` and this does not.